src/trees/FunctionTree.cpp

```diff
@@ -162,6 +162,25 @@ double FunctionTree<D>::evalf(const double *r) {
     return result;
 }
 
+
+template<int D>
```

Comment:
I think you can get away with one implementation for the three cases using `std::initializer_list` along these lines (haven't checked if this works)
```
template <int D>
double FunctionTree<D>::evalf(const std::initializer_list<double> & args) const {
   // Check length of argument list
   static_assert(args.size() == D, "More arguments than dimensions passed");
   return this->evalf(args);
}
```
This might also possibly simplify the binding of `FunctionTree`.